src/custom_instance_manager.ts

```diff
@@ -0,0 +1,112 @@
+import { exec } from 'child_process';
+import { InstanceGroup } from './instance_group';
+import { Context } from './context';
+
+function makeRandomString(length: number) {
+    let result = '';
+    const characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
+    const charactersLength = characters.length;
+    for (let i = 0; i < length; i++) {
+        result += characters.charAt(Math.floor(Math.random() * charactersLength));
+    }
+    return result;
+}
+
+export interface CustomInstanceManagerOptions {
+    isDryRun: boolean;
+}
+
+export default class CustomInstanceManager {
+    private isDryRun: boolean;
+
+    constructor(options: CustomInstanceManagerOptions) {
+        this.isDryRun = options.isDryRun;
+
+        this.launchInstances = this.launchInstances.bind(this);
+    }
+
+    async launchInstances(
+        ctx: Context,
+        group: InstanceGroup,
+        groupCurrentCount: number,
+        quantity: number,
+    ): Promise<Array<string | boolean>> {
+        ctx.logger.info(`[custom] Launching a batch of ${quantity} instances in group ${group.name}`);
+
+        const indexes: Array<number> = [];
+        for (let i = 0; i < quantity; i++) {
+            indexes.push(i);
+        }
+
+        const result = await Promise.all(
+            indexes.map(async (index) => {
+                return this.launchInstance(ctx, index, group);
+            }),
+        );
+        ctx.logger.info(`Finished launching all the instances in group ${group.name}`);
+
+        return result;
+    }
+
+    async launchInstance(ctx: Context, index: number, group: InstanceGroup): Promise<string | boolean> {
+        const groupName = group.name;
+        const groupInstanceConfigurationId = group.instanceConfigurationId;
+
+        const displayName = groupName + '-' + makeRandomString(5);
+
+        ctx.logger.info(`[custom] Launching instance number ${index + 1} in group ${groupName} with properties`, {
+            groupName,
+            displayName,
+            groupInstanceConfigurationId,
+        });
+
+        if (this.isDryRun) {
+            ctx.logger.info(`[custom] Dry run enabled, skipping the instance number ${index + 1} launch`);
+            return true;
+        }
+        try {
+            const launchResponse = await this.execLaunch({ displayName, groupName, region: group.region });
+            ctx.logger.info(
+                `[custom] Got launch response for instance number ${
+                    index + 1
+                } in group ${groupName}: ${launchResponse}`,
+            );
+
+            return launchResponse;
+        } catch (err) {
+            ctx.logger.error(
+                `[custom] Failed launching instance number ${index + 1} in group ${groupName} with err ${err}`,
+                { err },
+            );
+            return false;
+        }
+    }
+
+    async getInstances(): Promise<[]> {
+        return [];
```

Comment:
This would eventually keep a record of existing instances launched by the launcher and return them, I presume?

Reply:
Thanks for pointing it out.
I was originally not sure that would even be needed as this function is called by `reportUntrackedInstances`.
But from what I understand after another look, it indeed needs to return all instances.

Reply:
it looks like I can't really fake it as it needs to return status from the instance.